cmd/incusd/certificates.go

```diff
@@ -574,19 +574,24 @@ func certificatesPost(d *Daemon, r *http.Request) response.Response {
 					return response.Forbidden(fmt.Errorf("No matching certificate add operation found"))
 				}
 
-				tokenReq, ok := joinOp.Metadata["request"].(api.CertificatesPost)
-				if !ok {
-					return response.InternalError(fmt.Errorf("Bad certificate add operation data"))
-				}
-
 				// Create a new request from the token data as the user isn't allowed to override anything.
-				req = api.CertificatesPost{
-					CertificatePut: api.CertificatePut{
-						Name:       tokenReq.Name,
-						Type:       tokenReq.Type,
-						Restricted: tokenReq.Restricted,
-						Projects:   tokenReq.Projects,
-					},
+				req = api.CertificatesPost{}
+				switch tokenReq := joinOp.Metadata["request"].(type) {
+				case api.CertificatesPost:
+					req.Name = tokenReq.Name
+					req.Type = tokenReq.Type
+					req.Restricted = tokenReq.Restricted
+					req.Projects = tokenReq.Projects
+				case map[string]any:
+					req.Name = tokenReq["name"].(string)
+					req.Type = tokenReq["type"].(string)
+					req.Restricted = tokenReq["restricted"].(bool)
+					for _, project := range tokenReq["projects"].([]any) {
+						req.Projects = append(req.Projects, project.(string))
+					}
```

Comment:
Is there any possibility (user controlled) for any of those fields to be of different types?

Just want to make sure there's no way for someone to provide a bogus join token and crash Incus as a result.

Reply:
I don't think it's possible because the operation is retrieved from the node where the token has been created. `tokenReq` is just the result of the serialization of an `api.CertificatesPost` from the node where the token was issued. But I understand your concern so I can check all fields if you want.